dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/IdSchemeTests.java

```diff
@@ -0,0 +1,242 @@
+            Arguments.arguments( "NAME" ),
+            Arguments.arguments( "ATTRIBUTE:" + ATTRIBUTE_ID ) );
+    }
+
+    @MethodSource( "provideIdSchemeArguments" )
+    @ParameterizedTest( name = "POST to /tracker?idScheme={0}" )
+    public void shouldImportTrackerProgramDataByIdScheme( String idScheme )
+    {
+        String teiId = new IdGenerator().generateUniqueId();
+
+        TestData data = new TestData( idScheme );
+
+        EventDataBuilder eventDataBuilder = new EventDataBuilder()
+            .setOu( data.getOrgUnit() )
+            .setProgramStage( data.getTrackerProgramStage() )
+            .addDataValue( data.getDataElement(), DataGenerator.randomString() )
+            .setAssignedUser( Constants.SUPER_USER_ID )
+            .setProgram( data.getTrackerProgram() );
+
+        EnrollmentDataBuilder enrollmentDataBuilder = new EnrollmentDataBuilder()
+            .setOu( data.getOrgUnit() )
+            .setProgram( data.getTrackerProgram() )
+            .addEvent( eventDataBuilder );
+
+        JsonObject payload = new TeiDataBuilder()
+            .setId( teiId )
+            .setTeiType( data.getTrackedEntityType() )
+            .setOu( data.getOrgUnit() )
+            .addAttribute( data.trackedEntityAttribute, DataGenerator.randomString() )
+            .addEnrollment( enrollmentDataBuilder )
+            .addRelationship( new RelationshipDataBuilder()
+                .setRelationshipType( data.relationshipType )
+                .setFromTrackedEntity( this.createTei() )
+                .setToTrackedEntity( teiId ) )
+            .array();
+
+        trackerActions.postAndGetJobReport( payload,
+            new QueryParamsBuilder()
+                .addAll( "async=false", "idScheme=" + idScheme ) )
+            .validateSuccessfulImport();
+    }
+
+    @MethodSource( "provideIdSchemeArguments" )
+    @ParameterizedTest( name = "POST to /tracker?idScheme={0}" )
+    public void shouldImportEventProgramDataByIdScheme( String idScheme )
+    {
+        TestData data = new TestData( idScheme );
+
+        JsonObject object = new EventDataBuilder()
+            .setOu( data.getOrgUnit() )
+            .setProgramStage( data.getEventProgramStage() )
+            .setProgram( data.getEventProgram() )
+            .setAttributeCategoryOptions( Arrays.asList( data.categoryOption ) )
+            .array();
+
+        trackerActions
+            .postAndGetJobReport( object, new QueryParamsBuilder().addAll( "async=false", "idScheme=" + idScheme ) )
+            .validateSuccessfulImport();
+    }
+
+    @Test
+    public void shouldImportEventsWithDifferentProgramSchemes()
+    {
+        String programIdScheme = "CODE";
+        String ouIdScheme = "NAME";
+        String programStageIdScheme = "ATTRIBUTE:" + ATTRIBUTE_ID;
+        TestData data = new TestData();
+
+        JsonObject event = new EventDataBuilder()
+            .setProgram( new TestData( programIdScheme ).getEventProgram() )
```

Comment:
I might be wrong 😋 In `TestData(IdScheme)` we are making ~ 10 GET requests. Since we call this constructor 3 times in this test and 1 time in each of the above 2 tests (each parameterized with 3 values). That's 10 GET requests * (3+2*3) ~ 90 GET requests just to get the data we have in `idSchemesMetadata.json`

Do we parse/get such identifiers from test fixtures in other tests or do we always go via the API?

Reply:
You are not wrong, that's a good observation :) I can add a cache for responses if you want, or parse the values from the file, but all the tests take 10s here, so I don't think it's too bad.

Reply:
I would prefer reading them from the JSON file 😄 10s does not sound too bad, but this will add up over time 😬